src/lib/pgen.rs

```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
         }
     }
 }
+
+#[derive(PartialEq, Eq, Debug)]
+/// Implementation of `LR1Item` which is used to build the state graph of a grammar. Consists of a
+/// left-hand-side `String`, a right-hand-side list of `Symbols` and an integer `dot`, denoting the
+/// current position of the parser inside of `rhs`.
+pub struct LR1Item {
+    pub lhs: String,
+    pub rhs: Vec<Symbol>,
+    pub dot: usize,
+    //pub la: HashSet<String>
+}
+
+// TODO add customised Eq, PartialEq (if needed)
+impl LR1Item {
+    /// Returns a new LR1Item
+    ///
+    /// # Example
+    ///
+    /// ```c
+    /// let item = LR1Item::new("S".to_string(), vec![nonterminal!("A".to_string())], 1);
+    /// ```
+    pub fn new(lhs: String, rhs: Vec<Symbol>, dot: usize) -> LR1Item {
+        LR1Item {lhs: lhs, rhs: rhs, dot: dot}
+    }
+
+    /// Returns a copy of the `Symbol` at the current position inside the `rhs`.
+    ///
+    /// # Example
+    ///
+    /// ```c
+    /// let item = LR1Item::new("S".to_string(), vec![nonterminal!("A".to_string()),
+    ///                                               terminal!("a".to_string())], 1);
+    /// let n = item.next(); // returns terminal!("a")
+    /// ```
+    pub fn next(&self) -> Option<Symbol>{
+        if self.dot < self.rhs.len() {
+            let ref sym = self.rhs[self.dot];
+            Some(sym.clone())
+        }
+        else {
+            None
+        }
+    }
+}
+
+impl Hash for LR1Item {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // we don't need to hash the lookahead as LR1Items with the same lookahead can be merged
+        self.lhs.hash(state);
+        self.rhs.hash(state);
+        self.dot.hash(state);
+    }
+}
+
+/// Calculates the closure of a HashMap containing LR1Items
+pub fn closure1(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>,
+                closure: &mut HashMap<LR1Item, HashSet<String>>) {
+    let mut changed;
+    loop {
+        changed = false;
```

Comment:
Given that `mut` isn't used outside of the loop, maybe write `loop { let mut changed=false; ...}`? and delete the definition outside of the loop?